src/Conversion/ONNXToKrnl/NN/Pooling.cpp

```diff
@@ -30,288 +44,519 @@ Value emitScalarOpFor<ONNXMaxPoolSingleOutOp>(
-      Value identity = getIdentityValue<ONNXMaxPoolSingleOutOp>(
-          rewriter, loc, resultElementType);
-      rewriter.create<StoreOp>(loc, identity, alloc, resultIndices);
-
-      // 2.2 Define inner loops.
-      int nInnerLoops = kernelShape.size();
-      BuildKrnlLoop innerLoops(rewriter, loc, nInnerLoops);
-      innerLoops.createDefineAndOptimizeOp();
-      //   for Kx = 0 .. KX
-      for (int i = 0; i < nInnerLoops; ++i)
-        innerLoops.pushBounds(0, kernelShape[i]);
-
-      // 2.3 Emit inner loop nest.
-      innerLoops.createIterateOp();
-      rewriter.setInsertionPointToStart(innerLoops.getIterateBlock());
-      {
-        // 3. Emit inner loop body
-        // t = D[n][c][s1 * r1 + k1 * d1][s2 * r2 + k2 * d2];
-        // R[n][c][r1][r2] = max(R[n][c][r1][r2], t);
-
-        // 3.1 Prepare indices for accesing the data tensor.
-        SmallVector<Value, 4> dataIndices;
-        // 3.1.1 Batch indices: n, c
-        for (int i = 0; i < batchRank; ++i)
-          dataIndices.emplace_back(outerLoops.getInductionVar(i));
-        // 3.1.2 Insert spatial indices: sX * rX + kX * dX
-        for (int i = batchRank; i < nOuterLoops; ++i) {
-          // Get index along the inner loop's induction variables.
-          // It is used to obtain kernel/pad/stride/dilation index.
-          int j = i - batchRank;
-
-          Value spatialIndex = outerLoops.getInductionVar(i);
-          // If strides are present (not default) then emit the correct access
-          // index.
-          // sX *= rX
-          if (strides[i - batchRank] > 1) {
-            auto strideIndex = emitConstantOp(
-                rewriter, loc, rewriter.getIndexType(), strides[j]);
-            spatialIndex = rewriter.create<MulIOp>(
-                loc, strideIndex, outerLoops.getInductionVar(i));
+      // 2. Emit the body of the output loop nest, which applies a pooling
+      // window to a region in the input, producing one output pixel.
+      SmallVector<Value, 4> outputIndices;
+      for (int i = 0; i < outputShape.size(); ++i)
+        outputIndices.emplace_back(outputLoops.getInductionVar(i));
+
+      // 2.1 Emit: output[n][c][ho][wo] = identity
+      rewriter.create<StoreOp>(loc, identity, alloc, outputIndices);
+
+      // 2.2 Emit affine maps which express the lower and upper bounds for the
+      // pooling window's dimensions.
+      // The pooling window can be smaller than the kernel when slicing it over
+      // the border edges. Thus, we will compute the start and end indices for
+      // each dimension as follows.
+      //   firstValidH = ceil(float(ptH / dH)) * dH - ptH
+      //   startH = max(firstValidH, ho * sH - ptH)
+      //   endH = min(H, ho * sH + (kH - 1) * dH  + 1 - pbH)
+      //   hDim = round(float(endH - startH) / float(dH))
+
+      // Prepare induction variables and constants as arguments for the affine
+      // maps.
+      SmallVector<SmallVector<Value, 4>, 4> IVsAndConstants;
```

Comment:
I would really appreciate if IVsAndConstants can be represented using nested structs to make the semantic meaning of both data structure obvious for readers.